crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,160 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "crypto/s2n_ecc_evp.h"
+
+#include <openssl/ecdh.h>
+#include <openssl/evp.h>
+#include <stdint.h>
+
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_mem.h"
+#include "utils/s2n_safety.h"
+
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY *, EVP_PKEY_free);
+DEFINE_POINTER_CLEANUP_FUNC(EVP_PKEY_CTX *, EVP_PKEY_CTX_free);
+
+/* IANA values can be found here:
+ * https://tools.ietf.org/html/rfc8446#appendix-B.3.1.4 */
+/* Share sizes are described here:
+ * https://tools.ietf.org/html/rfc8446#section-4.2.8.2 and include the extra
+ * "legacy_form" byte */
+#if S2N_IS_X25519_SUPPORTED
+const struct s2n_ecc_named_curve s2n_ecc_curve_x25519 = {
+    .iana_id = TLS_EC_CURVE_ECDH_X25519, 
+    .libcrypto_nid = NID_X25519, 
+    .name = "x25519", 
+    .share_size = 32
+};
+#endif
+
+/* IANA values can be found here:
+ * https://tools.ietf.org/html/rfc8446#appendix-B.3.1.4 */
+/* Share sizes are described here:
+ * https://tools.ietf.org/html/rfc8446#section-4.2.8.2 and include the extra
+ * "legacy_form" byte */
+const struct s2n_ecc_named_curve *const s2n_ecc_evp_supported_curves[] = {
+    &s2n_ecc_curve_secp256r1,
+    &s2n_ecc_curve_secp384r1,
+#if S2N_IS_X25519_SUPPORTED
+    &s2n_ecc_curve_x25519,
+#endif
+};
+
+#if S2N_IS_X25519_SUPPORTED
+static int s2n_ecc_evp_generate_key_x25519(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+#endif
+
+static int s2n_ecc_evp_generate_key_nist_curves(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+static int s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey);
+static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key, EVP_PKEY *peer_public, struct s2n_blob *shared_secret);
+
+#if S2N_IS_X25519_SUPPORTED
+static int s2n_ecc_evp_generate_key_x25519(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey) {
+
+    DEFER_CLEANUP(EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(named_curve->libcrypto_nid, NULL),
+                  EVP_PKEY_CTX_free_pointer);
+    S2N_ERROR_IF(pctx == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    GUARD_OSSL(EVP_PKEY_keygen_init(pctx), S2N_ERR_ECDHE_GEN_KEY);
+    GUARD_OSSL(EVP_PKEY_keygen(pctx, evp_pkey), S2N_ERR_ECDHE_GEN_KEY);
+    S2N_ERROR_IF(evp_pkey == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    return 0;
+}
+#endif
+
+static int s2n_ecc_evp_generate_key_nist_curves(const struct s2n_ecc_named_curve *named_curve, EVP_PKEY **evp_pkey) {
+
+    DEFER_CLEANUP(EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL), EVP_PKEY_CTX_free_pointer);
+    S2N_ERROR_IF(pctx == NULL, S2N_ERR_ECDHE_GEN_KEY);
+
+    GUARD_OSSL(EVP_PKEY_paramgen_init(pctx), S2N_ERR_ECDHE_GEN_KEY);
+    GUARD_OSSL(EVP_PKEY_CTX_set_ec_paramgen_curve_nid(pctx, named_curve->libcrypto_nid), S2N_ERR_ECDHE_GEN_KEY);
+
+    DEFER_CLEANUP(EVP_PKEY *params = NULL, EVP_PKEY_free_pointer);
```

Comment:
NitPick: The `params` object can be safely reused (in a thread-safe manner). Thus you can probably achieve higher performance by only creating this object once per curve and re-using it.

Reply:
I see, will look into this option

Reply:
Recording offline discussion: If the server (or client) is generating many ephemeral ECDH keys on the same curve, given the `params` object can be re-used it only needs to generate a `params` object once for each curve and then can keep reusing it. This (should) be much faster than re-creating it every time.  The implementation would require initialization and cleanup of the params memory to a higher level call, perhaps by setting this up when s2n is initialized and clean it up when s2n is cleaned up.

Reply:
Deferring this optimization to a later change.